[addons][network] add new functions GetHostname(), IsLocalHost(...) and IsHostOnLAN(...) / minor cleanup

AlwinEsch opened the pull request:
## Description

Two new functions where good for inputstream addons.

Commit 1:
------------------------------
**[addons][network] add new functions**

 - GetHostname()
 - IsLocalHost(...)
 - IsHostOnLAN(...)

Commit 2:
------------------------------
**[addons][network] minor cleanups**

Some not needed doc lines are removed, use @ only for doxygen and use
new/delete for his interface structure.

Commit 3:
------------------------------
**[addons][network] separate "C" and "C++" parts of headers**

By them becomes the "C" part placed in own header to see better what language
is where and to make easier for other languages.

Commit 4:
------------------------------
**[addons][network] increase API version to 1.0.1**

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][network] minor cleanups

Some not needed doc lines are removed, use @ only for doxygen and use
new/delete for his interface structure.

AlwinEsch committed:
[addons][network] increase API version to 1.0.1

Rechi commented on xbmc/addons/interfaces/Network.cpp:
```diff
@@ -75,6 +76,39 @@ char* Interface_Network::get_ip_address(void* kodiBase)
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || hostname == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_Network::{} - invalid data (addon='{}', hostname='{}')", __FUNCTION__,
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi reviewed (commented):
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

phunkyfish reviewed (commented):
Looks good apart from the one clang format issue Rechi pointed out.

AlwinEsch committed:
[addons][network] separate "C" and "C++" parts of headers

By them becomes the "C" part placed in own header to see better what language
is where and to make easier for other languages.

AlwinEsch commented:
Have it updated and added also `IsHostOnLAN(...)` to there.

Further splitted like with PVR now the "C" part to own header.

AlwinEsch committed:
[addons][network] add new functions

- GetHostname()
- IsLocalHost(...)
- IsHostOnLAN(...)

To allow addon, e.g. inputstream to check his source on same PC.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Runtime with this, was OK:
```cpp
bool RuntimeTest_Network()
{
  bool ret;

  std::string ip = kodi::network::GetIPAddress();
  fprintf(stderr, "- kodi::network::GetIPAddress():               %s\n", ip.c_str());

  std::string hostname = kodi::network::GetHostname();
  fprintf(stderr, "- kodi::network::GetHostname():                %s\n", hostname.c_str());

  ret = kodi::network::IsLocalHost("127.0.0.1");
  fprintf(stderr, "- kodi::network::IsLocalHost(\"127.0.0.1\"):   %s\n", ret ? "yes" : "no");
  ret = kodi::network::IsLocalHost("localhost");
  fprintf(stderr, "- kodi::network::IsLocalHost(\"localhost\"):   %s\n", ret ? "yes" : "no");
  ret = kodi::network::IsLocalHost("192.168.0.1");
  fprintf(stderr, "- kodi::network::IsLocalHost(\"192.168.0.1\"): %s\n", ret ? "yes" : "no");
  ret = kodi::network::IsLocalHost("212.4.0.1");
  fprintf(stderr, "- kodi::network::IsLocalHost(\"212.4.0.1\"):   %s\n", ret ? "yes" : "no");

  std::string encodedUrl = kodi::network::URLEncode("François");
  fprintf(stderr, "kodi::network::URLEncode(\"François\")        '%s'\n", encodedUrl.c_str());

  std::string ipAddress;
  ret = kodi::network::DNSLookup("www.google.com", ipAddress);
  fprintf(stderr, "DNSLookup returned for www.google.com the IP '%s', call was %s\n", ipAddress.c_str(), ret ? "ok" : "failed");
  return true;
}
```
